### ISSUE TITLE
RFC: require alternative backends to pass SciPy's tests when "dispatching"?

### ISSUE THREAD

**Issue by lucascolley** [scipy/scipy#21031]
See https://github.com/scipy/scipy/pull/21012#issue-2365269519, https://github.com/scipy/scipy/pull/21012#pullrequestreview-2135129312.

Support for alternative backends which uses the method of "dispatching" (i.e. calling out to some known alternative backends based on the type of the input), like that in `special` currently, can be tested in one of (at least) two ways.

1. Leave the thorough testing of correctness to the alternative backends. Add some more simple tests that reassure us that the correct functions are being called. Direct any bug reports for corner-cases of an alternative backend to the library of that backend.
2. Convert all of SciPy's tests to be array-agnostic so that we can be sure that alternative backends are providing an accurate match of the SciPy functions. This will enable us to track where alternative backends do not match via skips/x-fails.

We don't yet have agreement on what to do at the moment.

---

My opinion is that (1) is absolutely fine while the behaviour is experimental and opt-in via an env variable. Any users who are going out of their way to help us test this experimental dispatching should be smart enough to send bug reports to the appropriate library, rather than to us. I think that our time is better spent on covering more of the codebase with support for alternative backends, than on converting tests to find issues in other libraries.

I think that (2) would be nice to have by the time we decide to make this behaviour the default, if only to help reduce the amount of bug reports, but hopefully the onus would still be on the other libraries to track down and fix the bugs in their implementations.

cc @ev-br @mdhaber @rgommers

**Comment by ev-br** [comment]
To clarify my comment https://github.com/scipy/scipy/pull/21012#pullrequestreview-2135129312:

1. Scipy tests fix what the SciPy functionality is.
2. When we add new functionality, it needs new tests.
3. When we add a new optional dependency, we check that existing tests pass, and skip/xfail as needed. If/when a dependency fixes the underlying upstream issue, we undo the skip/xfail. 

In my mind, adding dispatch to cupy/torch/jax/whatnot is more like adding new optional dependencies, not new functionality.

Also 

1. Adding a skip of a SciPy test with backend X does not make us responsible for X. It only signals what our guarantee for a SciPy user is.
2. If we have a skip/xfail, and there's a bug report on SciPy (and there will be!), we link to the skip and direct a user to the upstream library.
3. Adapting existing tests is not _that_ hard. For one example, see https://github.com/ev-br/scipy/commits/ndimage_array_api_tests/. Individual commits have `sed` incantations which helped a lot.

Concluding, I believe the right way is to first adapt the tests for the `xp_assert` machinery, and only then start adding dispatch. 
If there's appetite for improving tests themselves, great! This should come in a follow-up.

**Comment by mdhaber** [comment]
Thanks @ev-br. I think I understand your perspective, but you've mentioned that we don't *need* to add new tests or improve the existing tests; we *just* need to translate the old tests. This makes me think that the view of these new tests is to improve the coverage of the existing tests. That is a side benefit, but these new tests are not meant as improvements over the existing tests[^1]. Rather, the primary purpose is as a _substitute_ (temporary or not) for translating the existing tests.

I can see how these other backends can be considered optional dependencies, although in some cases they are a bit different because they aren't listed as optional dependencies in the usual places, and our code doesn't necessarily import them. (We only need to import them if the scipy-like functionality is in a totally different namespace.) In any case, if there is a strict rule that optional dependencies used to replicate existing functionality must pass all existing tests, I can see how your conclusions would follow. 

But I thought we have some flexibility as to how we test things, and that if we can create new tests that are just as strong as the existing tests with less work than translating the existing tests, it would be acceptable.

> Adapting existing tests is not that hard.

I've probably translated a few thousand lines of tests at this point, and in some cases I agree. There are even some advantages to doing so (e.g. update test style). But it is tedious, and it is usually the bulk of the work of doing these translations. It would be almost all of the work if we're really just dispatching to the other backend rather than translating SciPy's version. So I figure that we might as well save some author and review time if we can get just as strong or stronger tests of the other backend with less work.

[^1] Although they can be. By comparing SciPy's special functions against the other libraries, the tests in `special` have uncovered several bugs in SciPy (that are not covered by existing tests) and in the other library special functions.

**Comment by lucascolley** [comment]
> @lucascolley What do you think about changing
> 
> > Leave the thorough testing of correctness to the alternative backends. Add some more simple tests that reassure us that the correct functions are being called. Direct any bug reports for corner-cases of an alternative backend to the library of that backend.
> 
> I wouldn't say that the new property-based tests are not thorough; the thoroughness is tunable. In fact, we found that using Hypothesis for special was a bit too thorough because it uncovers all sorts of edge cases that shouldn't preclude SciPy dispatching to the function. In the short term, we can discover these sorts of bugs with Hypothesis locally; once the bugs are fixed, we can make the tests in CI stronger.
> 
> Maybe:
> 
> > Add property-based tests that demonstrate that the appropriate functions are being called correctly and produce results comparable to those of SciPy. Direct any bug reports for corner-cases of an alternative backend to the library of that backend.

Sure! My point was that the thoroughness isn't so important in this scenario at the minute (where we can effectively give alternative backends the benefit of the doubt until we think about seriously making this the default behaviour), but you're right that the substitute tests needn't be less thorough at all.

**Comment by rgommers** [comment]
I agree with most of what is written here. Lucas makes a good point about (2) being nicer than (1), but (1) probably being okay as a first step given everything is experimental. Evgeni wrote the "engineering summary" that I would have written more or less the same. And Matt does have a good point about the test translation being tedious, which has not been a showstopper so far but is not ideal.

* * *

> I wouldn't say that the new property-based tests are not thorough; the thoroughness is tunable. In fact, we found that using Hypothesis for special was a bit too thorough because it uncovers all sorts of edge cases that shouldn't preclude SciPy dispatching to the function. In the short term, we can discover these sorts of bugs with Hypothesis locally; once the bugs are fixed, we can make the tests in CI stronger.

Making much heavier use of Hypothesis is a separate discussion I think. Which isn't even directly related to support for other array libraries.

Hypothesis cannot really replace Pytest - it's too slow and, most importantly, not reproducible. This is not "being thorough", it is "not catering to regression testing of large libraries well". This is based on experience with `hypothesis` in NumPy, in SciPy (both use it sparingly), in `array-api-tests` (which uses it for 100% of tests), and various discussions over the years with hypothesis maintainers/experts.

Hypothesis can be great as an _extra_ tool. Especially when developing new code, it can find new bugs and make you think about corner cases that you otherwise would have missed. 

Fundamentally, we need two different things:
1. regression testing. key requirements: very stable, run fast, when there are failures they need to be understandable
2. fuzz testing. key requirement/goal: comprehensive testing of code, trying as hard as possible to find issues. (can be slow, typically non-reproducible).

Our primary requirement is (1), which is what `pytest` gives us. `hypothesis` property-based testing is basically a smart form of (2), and attempts to make it do (1) have not been very successful so far.

**Comment by mdhaber** [comment]
Note that property-based testing and Hypothesis don't need to be synonymous. The point is that the they property-based tests are written, you can use either "dumb" random numbers for basic regression testing or Hypothesis' "smart" numbers to find problems. In `special`, we are now performing fast, reproducible property-based tests without Hypothesis.

**Comment by rgommers** [comment]
Yes indeed, good point.

**Comment by ilayn** [comment]
For this, (I am feeling like I am parroting the same stuff), in my opinion, array API **must** provide all the testing tools including the *backend mocking suite* without installing JAX or CuPy or anything else as a dependency. You don't install a 4 GB database to test your ORM in the testing suite. Same should apply here. There should be a single thing and that should be the Array API testing suite. It is still in progress https://data-apis.org/array-api/latest/verification_test_suite.html Is there any reason why we cannot wait for that?

I don't know the background of these work for array API or why we are jumping on the gun before it matures (maybe quansight connection maybe something else whatever) but these tooling must be ready for the consumers of the array API. Not we going out of our ways to provide extensive work for it. Good intentions will not help in the end if array API cannot mature. We cannot be the catalyst for that.

If we need to test dispatching ourselves then something is not yet well-thought out in the specification. Not every adopting library (not array generating ones but consumers like SciPy) will have the luxury to do what we are doing here in SciPy. That means they won't be able to do these tests. And bad implementations will follow and eventually hurt the adoption. Adopting the array API is not clear at all, https://data-apis.org/array-api/latest/purpose_and_scope.html#how-to-adopt-this-api or I don't know where to look. 

Array API is supposed to be about array generating libraries providing a common infrastructure and not consumers testing out their dispatch mechanisms independently. The standard should have a definite tooling such that when things pass, they are compliant and they should stay compliant when new members join. This is not a new idea, serious standard tooling software and Bluetooth, USB, CE certification etc. standard do this. In other words, if a new mobile network operator joins say Vodafone, existing phone users should not update their firmware to use that operator because it does not understand + but needs 00 for international numbers. Currently what we are doing is, `if is_vodafone: ...replace('+', '00')` in the firmware of the phone. This feels quite awkward to me. Maybe I fundamentally don't get the API purpose. That is a very real possibility.

Take Matt's comprehensive dispatch mechanism in #21012 it is very clean but full of edge cases. If the array generators can't figure things out why should we be the arbiter? That is not an API standard that is SciPy going out of its way to accommodate all the array generating libraries under the sun. 

Moreover, say I have a library that I would like to support array API with. Am I going to copy SciPy's code to do the same? If I am going to copy it why do I not copy from the source which is the array API specification instead? If we are serious about the array API intentions we should work on the testing specs at the place where these are being generated. Otherwise I cannot see why SciPy is the experimentation ground for this API. I asked this multiple times here already so not to start that discussion once again.

In a world where unicorns exist, I should pip install `array_api` package (or whatever it is called), and it should come with the necessary clauses so that I don't need to do `is_cupy`, `is_numpy` and so on. It should do it internally because it knows the array adopters and all the relevant tooling. If a new contender joins the array generating libraries, I should be able to update the package version and things should keep working because I did my part by conforming to the standard. If we add conditionals to every case continuously, that's us maintaining array API as consumers. 

We keep on adding custom tooling into SciPy and I cannot see where it is going with all the assert discussions and dispatchers and custom decorators related to array API and how this will circle back to array API specifications in the end for the generic usage.

**Comment by lucascolley** [comment]
> For this, (I am feeling like I am parroting the same stuff), in my opinion, array API must provide all the testing tools including the backend mocking suite without installing JAX or CuPy or anything else as a dependency. You don't install a 4 GB database to test your ORM in the testing suite. Same should apply here. There should be a single thing and that should be the Array API testing suite. It is still in progress [data-apis.org/array-api/latest/verification_test_suite.html](https://data-apis.org/array-api/latest/verification_test_suite.html) Is there any reason why we cannot wait for that?

I think there are two separate concepts getting a little confused here. The quoted link is for array-provider libraries such as NumPy and PyTorch to verify that they are compatible with the standard. If we want to start accepting PyTorch arrays in SciPy, we need to verify not only that PyTorch is compatible with the standard, but that our implementations which use it are actually correct.

> If we need to test dispatching ourselves then something is not yet well-thought out in the specification. Not every adopting library (not array generating ones but consumers like SciPy) will have the luxury to do what we are doing here in SciPy. That means they won't be able to do these tests. And bad implementations will follow and eventually hurt the adoption. Adopting the array API is not clear at all, [data-apis.org/array-api/latest/purpose_and_scope.html#how-to-adopt-this-api](https://data-apis.org/array-api/latest/purpose_and_scope.html#how-to-adopt-this-api) or I don't know where to look.

The "dispatching" discussed in this issue is not about adopting the array API. SciPy is in a unique position in the ecosystem with a lot of compiled code, and a lot of "mirror" implementations for alternative backends for modules like `linalg` and `signal` which we can call out to. For the parts which the specification does cover (like the `fft` and `linalg` extensions), adopting support is rather simple. It's the other, more ambitious goals here that are challenging.

> Take Matt's comprehensive dispatch mechanism in https://github.com/scipy/scipy/pull/21012 it is very clean but full of edge cases. If the array generators can't figure things out why should we be the arbiter? That is not an API standard that is SciPy going out of its way to accommodate all the array generating libraries under the sun.

Indeed, this is work which goes rather further than what the standard is trying to help us with. I agree that we should focus on using the standard for our pure Python code for now, and focus on compiled code later (this is what the array types RFC said too). However, people clearly see the value in trying to call out to CuPy, JAX etc. in these places as it can bring GPU support, even if that means complex mechanisms and handling tricky edge cases.

>  say I have a library that I would like to support array API with. Am I going to copy SciPy's code to do the same?

this was discussed a little recently in gh-21054, and indeed the aim is to reach a point where any code which is reusable for other projects is upstream. It is just still very early days.

> In a world where unicorns exist, I should pip install array_api package (or whatever it is called), and it should come with the necessary clauses so that I don't need to do is_cupy, is_numpy and so on. It should do it internally because it knows the array adopters and all the relevant tooling. If a new contender joins the array generating libraries, I should be able to update the package version and things should keep working because I did my part by conforming to the standard.

This is pretty much the case for pure Python code, which is the scope that the standard is aiming to help with. We will still have `is_numpy` everywhere while we want to treat NumPy as the special, default backend. But `is_cupy` and so on are only needed for this "dispatching" which is an orthogonal effort.

I suppose the suggestion is more that there should be standard tooling which allows you to dispatch to alternative backends. That was also covered in the RFC and stated to be out of scope for now (see the "Dispatching mechanism" section of gh-18286).

---

The general sentiment that there is a strong need for standard testing tooling is important. I think that getting standardised versions of the assertions upstreamed would be very high value.

**Comment by fancidev** [comment]
> There should be a single thing and that should be the Array API testing suite. Is there any reason why we cannot wait for that?

I think the reality is that there will always be some backend that does not conform 100% to the Array API standard, but yet is popular enough that SciPy wants to make sure it works well with. That’s why I think the individual back end tests are going to stay.

> Good intentions will not help in the end if array API cannot mature. 

I guess having SciPy support the Array API could greatly facilitate adoption of the API. So it’s a positive feedback loop.

> In other words, if a new mobile network operator joins say Vodafone, existing phone users should not update their firmware to use that operator because it does not understand + but needs 00 for international numbers. Currently what we are doing is, `if is_vodafone: ...replace('+', '00')` in the firmware of the phone. 

I’m sure you’ve seen things like #ifdef __MSVC__ in a C code base. The situation is pretty much the same here. When there are just a few vendors, they tend to be powerful enough to not adhere to the standard, and library or even end users have to adapt. I think it’s getting better with C; maybe things will get better with Array API as the stakeholders evolve.
 
> That is not an API standard that is SciPy going out of its way to accommodate all the array generating libraries under the sun.

I guess the backend has to be popular enough to be supported. (Don’t ask me how popular is popular enough!)

> In a world where unicorns exist, I should pip install `array_api` package (or whatever it is called), and it should come with the necessary clauses so that I don't need to do `is_cupy`, `is_numpy` and so on. 

I thought there is an `array_api_compat` package, but I have the impression it only smooths some of the syntactic difference, but does nothing about the semantic differences. In addition, I’m under the impression that the semantic differences are impossible to systematically bridge. So perhaps unicorns don’t exist.

**Comment by ilayn** [comment]
> If we want to start accepting PyTorch arrays in SciPy, we need to verify not only that PyTorch is compatible with the standard, but that our implementations which use it are actually correct.

This I don't understand. We don't have a choice if we are participating in Array API. By definition it should be correct without us testing it. It is the very definition of Array API. If I need to check every backend then this is not a standard. It is SciPy vendoring Pytorch. So I am quite -1 on that. In other words,

- If I am only using it as an array object factory, its arrays should behave like NumPy or according to Array API standard. Otherwise it's not conforming
- Alternatively, if I am dispatching it to its own library, I must not and can not check its value correctness because it is not my business. It's pytorch's business. That's why @ev-br is saying and I am agreeing if I understood correctly. And to be perfectly honest, it is way early to do this at this stage while the standard is also figuring things out. There is no need to rush this at all.

> However, people clearly see the value in trying to call out to CuPy, JAX etc. in these places as it can bring GPU support, even if that means complex mechanisms and handling tricky edge cases.

Ok that's perfectly fine but once it is in, it is a pain to take things out. Why are we ... rushing ... for this. Rushing word is becoming my go to word for now apologies. 

> this was discussed a little recently in https://github.com/scipy/scipy/issues/21054, and indeed the aim is to reach a point where any code which is reusable for other projects is upstream. It is just still very early days.

Same thing. Why don't we add it to the array API side and not in SciPy then? Is there some blocking issue for not working this inside array API repo, then import it later, keeping SciPy code clean? 

> The general sentiment that there is a strong need for standard testing tooling is important. I think that getting standardised versions of the assertions upstreamed would be very high value.

Exactly, then why don't we do this on the Array API side and not here? 

<br>

> I think the reality is that there will always be some backend that does not conform 100% to the Array API standard, but yet is popular enough that SciPy wants to make sure it works well with. That’s why I think the individual back end tests are going to stay.

That is not our problem. If we adopt a standard it is a standard. Popularity won't cut this time. Historical examples have ample amount of evidence for this and we should not repeat that mistake, moreover, the array API itself should not recognize that vendor. This is the worst thing any body of standard can do. It's quite binary choice when things come to standards. 

> I’m sure you’ve seen things like #ifdef MSVC in a C code base. The situation is pretty much the same here.

That's exactly why it is called non-conforming C99 compiler. We do it because we are late to the party. If that compiler started today nobody would touch it. We only touch it because windows does not have any options. When it does (not holding my breath), MSVC is done. 

> I thought there is an array_api_compat package, but I have the impression it only smooths some of the syntactic difference, but does nothing about the semantic differences. In addition, I’m under the impression that the semantic differences are impossible to systematically bridge. So perhaps unicorns don’t exist.

I hear you but this is undermining the array API not my argument. It will either work, or go into the historical dustbin. There is no middle ground for standards. I didn't start this initiative and I am not against it but there is no kinda support for standards.

**Comment by mdhaber** [comment]
I'm still not sure if everyone is discussing the same thing here. Note that "array api" doesn't appear in the top post. That's because the array API standard itself is essentially irrelevant to this idea of "dispatching", which is for things that aren't covered by the standard. This issue is about accepting non-NumPy arrays and either a) by converting to NumPy, doing something with SciPy code, and converting back or b) calling the non-NumPy array's equivalent function and returning that result.

**Comment by lucascolley** [comment]
Agreed that there seem to be at least three separate topics being discussed here, so I would recommend creating new issues if there are specific points to be discussed. Since I was quoted, one reply:

---

> If we want to start accepting PyTorch arrays in SciPy, we need to verify not only that PyTorch is compatible with the standard, but that our implementations which use it are actually correct.
This I don't understand. We don't have a choice if we are participating in Array API. By definition it should be correct without us testing it. It is the very definition of Array API. If I need to check every backend then this is not a standard. It is SciPy vendoring Pytorch.

Sorry, I think my wording was unclear. We (SciPy) do not verify that PyTorch is compatible with the standard - but we (ecosystem) do (PyTorch + array-api-compat are verified using array-api-tests).

We (SciPy) test our own functions. We do this using array-api-strict in CI. *For added reassurance* (but not fundamentally necessarily), we also test with PyTorch and JAX as received via array-api-compat. The reasons for this are twofold:

1. It creates a positive feedback loop, meaning that the correctness of upstream libraries (array-api-compat, the array libraries themselves) are checked against SciPy (as well as their own tests). This is just a nice thing to do, but not necessary from SciPy's side. This is especially useful while support is still in heavy development.
2. array-api-strict has some [caveats](https://data-apis.org/array-api-strict/#caveats). Checking against the real-world implementations, as well as the mock strict backend, is practical as there may be coverage of the few corners array-api-strict might not hit.

**Comment by ilayn** [comment]
That's fine but then Array API linalg extension should not be there because there is no linalg backend that is not compiled in any shape in any other library. So I'm not sure what the intention is with those extensions. If you need eigenvalues and use linalg extension of the Array API you must dispatch no?

**Comment by mdhaber** [comment]
@ilayn to avoid mixing issues, let's take the example of special functions, for which there is no standard.

For some special functions, SciPy now accepts non-NumPy arrays as arguments, and returns the result of the function evaluation as an array of the original type.

```python3
import torch
from scipy import special
special.ndtr(torch.asarray([1., 2., 3.]))
# tensor([0.8413, 0.9772, 0.9987])
```

SciPy uses torch's implementation of `ndtr` to accomplish this. For other special functions, it converts the torch array to a NumPy array, uses SciPy's compiled implementation of `ndtr`, and converts the result back to a torch Tensor.

This is what is meant by "dispatching".

This is valuable to do because it allows one to write a library that uses special functions in a backend agnostic way. Just call SciPy functions, and they will "just work" regardless of the array type. (Of course there are caveats. That's the dream.)

To simplify, the question at hand is how do we test our "dispatching" to make sure that it is working as we intend it to.

a) Execute all of SciPy's existing tests `ndtr` with an array of each backend and ensure that they pass.
OR
b) Make up representative values of argument `x`, and compare the results of `ndtr(x)` and `ndtr(torch.asarray(x))`

**Comment by ilayn** [comment]
OK let me try to give a similar example why these are all intertwined and in my opinion. 

The reason why we want to have non-numpy arrays (at least for Python code) is that if they agree on certain operations in the standard, then we use those subset of operations in the standard, and all will go precisely as intended regardless of the array type. That's what I understand from the Array API parts of this discussion. 

But then, we accept the same non-numpy arrays for which there is a compiled code in the way and obviously we wrote them for NumPy arrays. So we can't support Array API natively and we check if there is a functionality already in the vendor of that object. In fact for some there are for `special` and some `linalg` some `signal` etc too. 

Here is where these concepts gets mixed up. You have been arguing that SciPy is a special case for this, and I am saying no this is a fundamental problem because there is not much choice other than compiled code when it comes to significant numerical code; if `networkx` wants to do something this is an issue, if scikit-learn wants it, again there is a need for dispatching. 

Let say you want the determinant of a "tensor". It can be a NumPy array, it can be torch tensor etc. Where do they want to send it to for scikit-learn? Do they want to send it to SciPy and let us figure it out for them by dispatching? Or do they want to dispatch that themselves? It is compiled code wherever you send it to or do it yourself. 

If it is the former then this is a fundamentally different problem. I am arguing as I think it is the latter that will happen. If I am wrong then it is all a different problem completely and don't know where we decided to be that keystone. Because then nobody needs to adopt the array API and SciPy is the only user. 

But if it is the latter, they must dispatch one way or the other. So from this argument I am saying dispatching is not independent from Array API adoption and dispatching code should be the part of the Array API. And then my arguments above follow.

**Comment by mdhaber** [comment]
> Where do they want to send it to for scikit-learn? Do they want to send it to SciPy and let us figure it out for them by dispatching? 

I assume the former - they can send to SciPy and let us figure it out.

**Comment by lucascolley** [comment]
There is a bit of writing about this in the standard docs: https://data-apis.org/array-api/2023.12/design_topics/C_API.html. It gives there the example of scikit-learn. The standard is honest in saying "at the moment, the standard won't help you where the modules/functions use Cython". It recommends adding support per submodule, which makes it easier to explain to users which functionality is supported. It also says:

> Longer term: specific support for particular array types (e.g. cupy.ndarray can be supported with Python-only code via cupy.ElementwiseKernel).

But the "longer term" is advising that such options are explored after reaching the limits of the standard.

> dispatching is not independent from Array API adoption and dispatching code should be the part of the Array API

I think this is mistaking the scope of the array API standard for an all-in-one solution for alternative backends. Indeed, consumer libraries would like a standardised way to dispatch to other libraries, but that is a separate discussion.

**Comment by lucascolley** [comment]
> Let say you want the determinant of a "tensor". It can be a NumPy array, it can be torch tensor etc. Where do they want to send it to for scikit-learn?

The determinant is standardised as `xp.linalg.det`. So if scikit-learn is happy to require that alternative backends implement the `linalg` extension, they will use that. This is the general pattern for replacing NumPy calls. If scikit-learn is currently calling SciPy, they will continue to call SciPy, either should work. If scikit-learn is currently implementing their own functionality based on calls to `np`, they will continue to do so with calls to `xp`. Hence I don't understand the claim that:

> nobody needs to adopt the array API and SciPy is the only user.

**Comment by ilayn** [comment]
> The determinant is standardised as xp.linalg.det. So if scikit-learn is happy to require that alternative backends implement the linalg extension, they will use that.

Rhetorically, then why are we dispatching? Let's call `xp.linalg.det` ourselves. We don't because we have our own `det`. Do you see the issue when a library has its own implementation for any extension-standardized function? Hence, you must dispatch to support everything or send it to a library that supports everything that is only SciPy at the moment. That is the claim.

I am not saying we should not do it. But as you mentioned a few times, many details are still deferred to future and I don't think this is wise move to do things all the way at this early stage. 

Positive feedback and all, I hear you. But optimism is not a specification as of now. And we are too invested in believing things will work out in the end. I somehow don't feel like it yet. It does not mean pessimism is the right thing to do. But caution is not the enemy of the adoption.

**Comment by mdhaber** [comment]
@ilayn would you like to open an issue about that? It may be worth reconsidering, but the fact is that we currently are dispatching, so this issue assumes that we do dispatching, and the question is about standards for testing.

**Comment by ilayn** [comment]
Sure, will do. Ping me if I forget about it please.

**Comment by ev-br** [comment]
> I don't know the background of these work for array API or why we are jumping on the gun before it matures (maybe quansight connection maybe something else whatever) 

Mind expanding on this @ilayn ? Would be great to clear up possible concerns.

**Comment by ilayn** [comment]
No concerns. It's not a controversial statement. Given how conservative we are typically towards any addition backend or codebase, adopting a standard that is at its infancy, must have come from somewhere else. I just don't know it. But I guess there were people involved. Not that I spent any thought on it anyways.

**Comment by ev-br** [comment]
> Given how conservative we are typically towards any addition backend or codebase, adopting a standard that is at its infancy, must have come from somewhere else. I just don't know it. But I guess there were people involved. Not that I spent any thought on it anyways.

In my mind (*), we are not adopting anything just yet. All array api tweaks require an explicit opt-in by an env variable and is clearly documented as experimental. Maybe we could/should make the "this is experimental, do not use in production" message more loud and clear in the docs. 

Again IMO and FWIW, the array api work is nor more and no less than hashing out the tooling around the array API standard and smoothing out rough edges of the standard itself. That it happens in and around scipy is just a consequence of the overlap between us who care about the standard and us scipy devs. Again, while I don't see a big problem here, am happy to keep discussing potential concerns and ways to address them.
 


(*) personal opinion; is neither discussed with nor endorsed by and does not necessarily represent views of my employer

**Comment by rgommers** [comment]
> I put some elbow grease to where my mouth was in [#21031 (comment)](https://github.com/scipy/scipy/issues/21031#issuecomment-2186180119), and here's a concrete example to evaluate: in #21091 converts `ndimage` tests. A short summary:

This looks like very nice work @ev-br. And well worth the effort, given you found 11 CuPy issues and can now be reasonably sure that everything else does match (to the extent that there are SciPy tests for something). I think it's a convincing data point for staying with the "convert tests" approach.

That to me also makes sense from an end user perspective. A user does not know, and should not have to care, how SciPy implements support for CuPy under the hood. All they see is that `scipy.xxx.func(a_cupy_array)` returns a cupy array. Could be done by calling CuPy for compiled code, or by using the array API standard, or by something completely different. Our guarantees, and hence our testing methods, should not depend on how it's implemented either.

Conclusion: I think the best decision here is to test other array libraries by converting tests to `xp`-style, like we've been doing more or less from the very first PR for array types support (gh-18668).

**Comment by lucascolley** [comment]
> Conclusion: I think the best decision here is to test other array libraries by converting tests to xp-style, like we've been doing more or less from the very first PR for array types support (https://github.com/scipy/scipy/pull/18668).

I agree, but I'm not sure that is enough to close this issue. I think we are all pretty much agreed that we want to ensure that functionality passes SciPy's tests before we make any "guarantees". But does that mean that we should not accept a PR which uses property-based tests instead to justify its correctness? I am undecided - on the one hand, that seems to be creating more work overall by delaying the inevitable. But on the other hand, accepting those PRs would likely allow us to unblock other array API work. For example, some of the `special` functions are being used in converted `stats` functions (for which the tests are being converted), despite the `special` tests themselves having not been converted yet.

**Comment by mdhaber** [comment]
Yeah, I think I would need to understand the metrics better to decide what is best. It took just a few minutes to add the names and number of arguments of ~50 new special functions, and Hypothesis founds discrepancies between CuPy and SciPy in a dozen of them (just for arguments in the range `(-10, 10`)`). It is fine with me to convert tests to xp-style - that's why we're doing it in `stats`; I just think property based tests like in special should be _allowed_ if they are faster to write.

**Comment by lucascolley** [comment]
> I think it's great to convert tests to xp-style in some cases - that's why we're doing it in stats. I just think property based tests like in special should be allowed if they are deemed faster to write and review - or worth any extra time for the additional test strength - by those who are doing the work.

Given that there hasn't been a response to this (which sounds reasonable to me), I'll say that I'm happy to review code based on either approach, with the view to eventually converting all of the tests down the line. I'll close this now but feel free to reopen!
